Mathlib/Tactic/SuccessIfFailWithMsg.lean

```diff
@@ -0,0 +1,57 @@
+/-
+Copyright (c) 2017 Mario Carneiro. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro, Simon Hudon, Sébastien Gouëzel, Scott Morrison, Thomas Murrills
+-/
+import Lean
+import Std.Util.TermUnsafe
+
+/-!
+# Success If Fail With Message
+
+This file implements a tactic that succeeds only if its argument fails with a specified message.
+
+It's mostly useful in tests, where we want to make sure that tactics fail in certain ways under
+circumstances.
+-/
+
+open Lean Elab Meta Tactic Syntax
+
+namespace Mathlib.Tactic
+
+/-- `success_if_fail_with_msg msg tacs` runs `tacs` and succeeds only if they fail with the message
+`msg`.
+
+`msg` can be any term that evaluates to an explicit `String`. -/
+syntax (name := successIfFailWithMsg) "success_if_fail_with_msg " term:max tacticSeq : tactic
+
+/- Note that we don't use the following in our definition of `success_if_fail_with_msg`, because we
+want to check for definitional equality to the Expr specified by the syntax `$msg:term` there
+instead. We also can provide more descriptive error messages in the elaboration case. -/
+
+/-- Evaluates `tacs` and succeeds only if `tacs` both fails and throws an error equal (as a string)
+to `msg`. -/
+def successIfFailWithMessage [Monad m] [MonadLiftT IO m] [MonadBacktrack s m] [MonadError m]
+    (msg : String) (tacs : m α) (ref : Option Syntax := none): m Unit := do
+  let s ← saveState
+  let err ←
+    try _ ← tacs; pure none
+    catch err => pure (some (← err.toMessageData.toString))
+  restoreState s
+  if let some err := err then
+    unless msg == err do
+      if let some ref := ref then
+        throwErrorAt ref "tactic '{ref}' failed, but got different error message:\n\n{err}"
+      else
+        throwError "tactic failed, but got different error message:\n\n{err}"
+  else
+    if let some ref := ref then
+      throwErrorAt ref "tactic '{ref}' succeeded, but was expected not to"
+    else
+      throwError "tactic succeeded, but was expected not to"
+
+elab_rules : tactic
+| `(tactic| success_if_fail_with_msg $msg:term $tacs:tacticSeq) =>
+  Term.withoutErrToSorry <| withoutRecover do
+    let msg ← unsafe Term.evalTerm String (.const ``String []) msg
+    successIfFailWithMessage msg (evalTactic tacs) tacs
```

Comment:
```suggestion
    successIfFailWithMessage msg (evalTacticSeq tacs) tacs
```
I don't really understand the difference, but I think this is what you should use (can you test where you give a sequence of tactics as argument?)

Reply:
Oddly, they both seem to work with tactic sequences. I've just inquired further about the difference [on zulip](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/Difference.20between.20.60evalTactic.60.20and.20.60evalTacticSeq.60.3F).

To be safe, I've used `evalTacticSeq` for now; some tests already took a tactic sequence as argument (`trivial; trivial`), but I've also added one to test multiline tactic sequences, just in case.